x/upgrade/internal/types/storeloader.go

```diff
@@ -0,0 +1,129 @@
+package types
+
+import (
+	"encoding/json"
+	"fmt"
+	"path/filepath"
+
+	"github.com/spf13/viper"
+
+	"io/ioutil"
+	"os"
+
+	"github.com/cosmos/cosmos-sdk/baseapp"
+	"github.com/cosmos/cosmos-sdk/client/flags"
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// StoreLoaderWithUpgrade is used to prepare baseapp with a fixed StoreLoader
+// pattern. This is useful in test cases, or with custom upgrade loading logic.
+func StoreLoaderWithUpgrade(storeUpgrades *storetypes.StoreUpgrades) baseapp.StoreLoader {
+	return func(ms sdk.CommitMultiStore) error {
+		// TODO cleanup viper
+		home := viper.GetString(flags.FlagHome)
+		upgradeInfoPath := filepath.Join(home, "upgrade-info.json")
+
+		var lastBlockHeight = ms.LastCommitID().Version
+		upgrades, err := GetUpgradeInfoDataFromFile(upgradeInfoPath)
+
+		// There should not be any error in reading upgrade info from filesystem
+		if err != nil {
+			return err
+		}
+
+		// Check if the current commit version and upgrade height matches
+		if (len(upgrades.StoreUpgrades.Renamed) > 0 || len(upgrades.StoreUpgrades.Deleted) > 0) &&
```

Comment:
This looks like similar logic (but cleaned up) from the original upgrade handler.

Big issue here, is that we now never write the store upgrades from the pre-upgrade binary (design was that they would  be in the proposal), but use them from the post-upgrade binary (written in code). I understand the reason behind this change, however, here we check value from file.

I would actually remove the `StoreUpgrades` field form UpgradeInfo as it is confusing and not used elsewhere in the code (I think). And you should be checking eg. `len(storeUpgrades.Renamed)` here. I see you do use the proper `storeUpgrades` as an argument to `ms.LoadLatestVersionAndUpgrade`, so it seems that you just missed to upgrade the if statement.

Reply:
Yeah,sorry missed this. Changed in next commit